src/tools/execution-results.h

```diff
@@ -121,20 +121,22 @@ struct ExecutionResults {
   }
 
   bool areEqual(Literal a, Literal b) {
-    if (a.type != b.type) {
-      std::cout << "types not identical! " << a << " != " << b << '\n';
-      return false;
-    }
-    if (a.type.isRef()) {
+    if (a.type.isRef() && b.type.isRef()) {
       // Don't compare references - only their types. There are several issues
       // here that we can't fully handle, see
       // https://github.com/WebAssembly/binaryen/issues/3378, but the core issue
       // is that we are comparing results between two separate wasm modules (and
       // a separate instance of each) - we can't really identify an identical
       // reference between such things. We can only compare things structurally,
       // for which we compare the types.
+      // Another issue is that the same module, when passed through --roundtrip,
+      // will end up with a type that does not compare equally.
       return true;
     }
+    if (a.type != b.type) {
```

Comment:
An alternative would be to not special-case ref types, but to change the condition to `a.type != b.type && (a.type < b.type || b.type < a.type)`, which will detect structurally equivalent types correctly.

Reply:
But would that work with recursive types, for which we create a new one when roundtripping which is the issue here?

Reply:
Yes. `operator<` compares structure, not identity. (The bug it has where it can compare two types as mutually less than each other should not affect this use case.)

Reply:
Oh right, the problem is just of comparing equality via identity... changed as you suggested.